src/controllers/legacycontrollersettings.cpp

```diff
@@ -274,3 +291,143 @@ QWidget* LegacyControllerEnumSetting::buildInputWidget(QWidget* pParent) {
+    return element.hasAttribute("type") &&
+            QString::compare(element.attribute("type"),
+                    "color",
+                    Qt::CaseInsensitive) == 0;
+}
+
+void LegacyControllerColorSetting::parse(const QString& in, bool* ok) {
+    if (ok != nullptr) {
+        *ok = false;
+    }
+    reset();
+    save();
+
+    m_savedValue = QColor(in);
+    if (!m_editedValue.isValid()) {
+        *ok = false;
+        return;
+    }
+    m_editedValue = m_savedValue;
+}
+
+QWidget* LegacyControllerColorSetting::buildInputWidget(QWidget* pParent) {
+    auto* pPushButton = new QPushButton(tr("Change color"), pParent);
+
+    auto setColorIcon = [pPushButton](const QColor& color) {
+        QPixmap icon(24, 24);
+        QPainter painter(&icon);
+        painter.fillRect(0, 0, 24, 24, color);
+        pPushButton->setIcon(QIcon(icon));
+    };
+
+    connect(this,
+            &AbstractLegacyControllerSetting::valueReset,
+            pPushButton,
+            [this, pPushButton, setColorIcon]() {
+                if (m_editedValue.isValid()) {
+                    setColorIcon(m_editedValue);
+                } else {
+                    pPushButton->setIcon(QIcon());
+                }
+            });
+
+    connect(pPushButton, &QPushButton::clicked, this, [this, pPushButton, setColorIcon](bool) {
+        auto color = QColorDialog::getColor(m_editedValue, pPushButton, tr("Choose a new color"));
+        if (color.isValid()) {
+            m_editedValue = color;
+            setColorIcon(color);
+            emit changed();
+        }
+    });
+
+    if (m_savedValue.isValid()) {
+        setColorIcon(m_savedValue);
+    }
+
+    return pPushButton;
+}
+
+LegacyControllerFileSetting::LegacyControllerFileSetting(
+        const QDomElement& element)
+        : AbstractLegacyControllerSetting(element) {
+    m_defaultValue = QFileInfo(element.attribute("default"));
+    m_fileFilter = element.attribute("pattern");
+    m_savedValue = m_defaultValue;
+    m_editedValue = m_defaultValue;
+    reset();
+    save();
+}
+
+bool LegacyControllerFileSetting::match(const QDomElement& element) {
+    return element.hasAttribute("type") &&
+            QString::compare(element.attribute("type"),
+                    "file",
+                    Qt::CaseInsensitive) == 0;
+}
+
+void LegacyControllerFileSetting::parse(const QString& in, bool* ok) {
+    if (ok != nullptr) {
+        *ok = false;
+    }
+    reset();
+    save();
+
+    m_editedValue = QFileInfo(in);
+    if (!m_editedValue.exists()) {
+        *ok = false;
+        return;
+    }
+    m_savedValue = m_editedValue;
+}
+
+QWidget* LegacyControllerFileSetting::buildInputWidget(QWidget* pParent) {
+    auto* pWidget = new QWidget(pParent);
+    pWidget->setLayout(new QHBoxLayout);
+    auto* pPushButton = new QPushButton(tr("Browse..."), pWidget);
+    auto* pLabel = new QLabel(QStringLiteral("<i>%1</i>").arg(tr("No file selected")), pWidget);
```

Comment:
wdyt of this to make sure the format string doesn't get out of sync?
```suggestion
    auto* pLabel = new QLabel(QStringLiteral("<i>%1</i>").arg(tr("No file selected")), pWidget);
    auto setLabelText = [pLabel](QString&& text) {
        pLabel->setText(QStringLiteral("<i>%1</i>").arg(text));
    };
```